Create Dockerfile

Bobarshad opened the pull request:
Dockerfile to great new Windows Container to build tensorflow

Bobarshad committed:
Create Dockerfile

Dockerfile to great new Windows Container to build tensorflow

Bobarshad committed:
Update Dockerfile

Adding Dockerfile details

gbaned commented:
Hi @Bobarshad Can you please sign CLA. Thank you!

MichaelHudgins commented on tensorflow/tools/tf_windows_build_dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,132 @@
+
+RUN netsh interface ipv4 set subinterface \"vEthernet (Ethernet)\" mtu=1460 store=persistent
+
+RUN md C:\TEMP
+RUN md C:\TMP
```
Are we able to get rid of one of these temp directories?

Bobarshad replied:
C:\TEMP was removed and tested and it was found that C:\TMP is empty when running Docker.

MichaelHudgins commented on tensorflow/tools/tf_windows_build_dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,132 @@
+RUN md C:\TEMP
+RUN md C:\TMP
+
+RUN (New-Object Net.WebClient).DownloadFile('https://aka.ms/vs/16/release/vs_community.exe', 'C:\TEMP\vs_community.exe');
```
To reduce image size can we clean up downloaded files at the end of the file?

Bobarshad replied:
It was tested the C:\TMP is empty when running Docker.

MichaelHudgins commented on tensorflow/tools/tf_windows_build_dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,132 @@
+SHELL ["powershell.exe", "-ExecutionPolicy", "Bypass", "-Command", "$ErrorActionPreference='Stop'; $ProgressPreference='SilentlyContinue'; $VerbosePreference = 'Continue';"]
+
+# Install 7-Zip and add it to the path.
+RUN (New-Object Net.WebClient).DownloadFile('https://www.7-zip.org/a/7z2201-x64.msi', '7z.msi'); \
```
Can we update this?

Bobarshad replied:
The new version available is 7-Zip 23.00 (beta). Would it be advisable to proceed with installing this beta version?

MichaelHudgins commented on tensorflow/tools/tf_windows_build_dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,132 @@
+# Install bazelisk
+RUN md C:\tools\bazel
+RUN (New-Object Net.WebClient).DownloadFile( \
+         'https://github.com/bazelbuild/bazelisk/releases/download/v1.11.0/bazelisk-windows-amd64.exe', \
```
Can we update bazelisk version?

Bobarshad replied:
bazelisk was updated to v1.16.0

Bobarshad commented:
I will test all the above suggestions and update the Dockerfile accordingly.

Bobarshad commented:
> Hi @Bobarshad Can you please sign CLA. Thank you!

Done

Bobarshad commented on tensorflow/tools/tf_windows_build_dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,132 @@
+# Install the pip packages
+RUN python -m pip install --ignore-installed --force-reinstall --upgrade \
+    setuptools packaging\
+    pyreadline tf-estimator-nightly tb-nightly backports.weakref==1.0rc1 \
```
Do we have different requirements file for different Python versions? these are just works for Payton 3.9.7.

mihaimaruseac replied:
+1 to moving these to requirements file

Bobarshad replied:
Done.

Bobarshad commented on tensorflow/tools/tf_windows_build_dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,132 @@
+RUN (New-Object Net.WebClient).DownloadFile('https://aka.ms/vs/16/release/vs_community.exe', 'C:\TEMP\vs_community.exe');
+# Install Visual Studio 2019 Build Tools.
+SHELL ["cmd", "/S", "/C"]
+RUN C:\TEMP\vs_community.exe \
```
I have tested it works.

angerson replied:
Please give more details... what did you do to test it?

Bobarshad replied:
I have explained all the steps in [README.md](https://github.com/tensorflow/tensorflow/blob/317846107a8da70fac317dd16ecce0f5c7eda6a8/tensorflow/tools/tf_windows_build_dockerfiles/README.md)

Commit:
Rename Dockerfile to Dockerfile_py3.9

Commit:
Create README.md

Commit:
Create Dockerfile_py3.11

Commit:
Update README.md

Bobarshad commented on tensorflow/tools/tf_windows_build_dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,132 @@
+# Install pacman packages.
+RUN C:\tools\msys64\usr\bin\bash.exe -lc 'pacman --noconfirm -Syy git curl zip unzip patch'
+
+# Install Python 3.9.7
```
Yes, we have two Dockerfiles one for Python 3.9 and one for 3.11 now.

Commit:
Update Dockerfile_py3.9

Commit:
Update Dockerfile_py3.11

Commit:
Update README.md

mihaimaruseac commented on tensorflow/tools/tf_windows_build_dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,132 @@
+    [Environment]::SetEnvironmentVariable('PATH', $env:PATH, 'Machine');
+
+
+# Disable signature checking on pacman because we cannot initialize the keyring
```
+1 to not disabling signature checking.

There is a potential alternative to record all packages together with their checksum in a file and download them, check the checksum and only install after, but:
- this would reimplement pacman
- this would require us to maintain this list manually, so slowing down iterations

mihaimaruseac commented on tensorflow/tools/tf_windows_build_dockerfiles/Dockerfile_py3.9:
```diff
@@ -0,0 +1,111 @@
+# This Dockerfile creates an image that:
```
Since you have 2 files that should mostly be identical, any thoughts on removing duplication? We want to prevent edits to one file which are accidentally not propagated to the other one

MichaelHudgins replied:
Agreed this should follow how linux has a single file for the different versions.

Bobarshad replied:
Done

mihaimaruseac commented on tensorflow/tools/tf_windows_build_dockerfiles/README.md:
```diff
@@ -0,0 +1,23 @@
+**To build a Windows container locally, execute these commands. These commands have been tested on Windows PowerShell when run as an administrator:**
```
I'm not sure this is needed?

mihaimaruseac replied:
Yes, but we don't need them in the repo, they can be just a comment in the PR

Bobarshad replied:
Done

mihaimaruseac commented on tensorflow/tools/tf_windows_build_dockerfiles/README.md:
````diff
@@ -0,0 +1,23 @@
+
+**To build TensorFlow pip_package:**
+```
+#Change to tensorlflow root directory
````
Typo, 
```suggestion
#Change to tensorflow root directory
```

Commit:
Update Dockerfile_py3.9

clang upgrade to 16.0.4

Commit:
Update Dockerfile_py3.11

clang upgrade to 16.0.4

Bobarshad commented on tensorflow/tools/tf_windows_build_dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,132 @@
+
+# Install clang.
+RUN (New-Object Net.WebClient).DownloadFile( \
+         'https://github.com/llvm/llvm-project/releases/download/llvmorg-14.0.0/LLVM-14.0.0-win64.exe', \
```
clang upgrade to 16.0.4 for both python 3.9 and 3.11

MichaelHudgins commented on tensorflow/tools/tf_windows_build_dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,132 @@
+    Remove-Item .\7z.msi;
+
+# Install msys2, and add some extra tools.
+RUN (New-Object Net.WebClient).DownloadFile( \
```
Agreed, this is one of the things that will make both the dockerfiles more readable and what is installed more clear.

Bobarshad replied:
The 'Install msys2' and Visual Studio RUN commands have been moved into a separate ps1 script and tested. This process will be followed for all other long RUN commands.

Bobarshad replied:
All lengthy commands have been relocated to separate scripts.

angerson commented on tensorflow/tools/tf_windows_build_dockerfiles/README.md:
````diff
@@ -0,0 +1,23 @@
+docker run --name tf -itd --rm -v C:\tf\tensorflow:C:\workspace -v C:\tf\tmp:C:\tmp -e TEST_TMPDIR=C:\tmp -w C:\workspace win-docker bash
+```
+
+**To build TensorFlow pip_package:**
````
This seems to indicate that the containers are only useful for building TensorFlow. Am I mistaken in that the goal for these containers is to replace the complete Windows build+test environment (CPU: build + test;  GPU: just build) similar to the way the newer Linux containers did? You'll need to make sure they can also replicate our current Windows test suites, plus the libtensorflow builds.

Bobarshad replied:
It is changed.

angerson commented on tensorflow/tools/tf_windows_build_dockerfiles/README.md:
```diff
@@ -0,0 +1,23 @@
+md C:\tf
+md C:\tf\tmp
+cd C:\tf
+git clone https://github.com/Bobarshad/tensorflow.git
```
This should be tensorflow/tensorflow, not your personal repository.

Bobarshad replied:
Done

angerson commented:
Can you update the description for this PR to describe more about what you're trying to accomplish here and how it's going?

The container in its current form is quite far away from our goals (refer to the internal bug that spurred this work, cv-101). It hasn't been demonstrated that it will work with RBE, with tests, or for anything more than building the TF package on an unspecified Windows system. What's your assessment of how the work is going? What do you see happening next?

Commit:
Update README.md

Commit:
Update and rename Dockerfile_py3.11 to Dockerfile

Commit:
Delete Dockerfile_py3.9

Commit:
Create devel.requirements.txt

Commit:
Update README.md

Commit:
Update README.md

 Change to : git clone https://github.com/tensorflow/tensorflow.git

mihaimaruseac commented on tensorflow/tools/tf_windows_build_dockerfiles/README.md:
````diff
@@ -0,0 +1,36 @@
+**Starting docker:**
+
+```
+#Change to tensorlflow root directory
````
You still have this typo, even though I have a comment + suggested edit before.

Bobarshad replied:
Fixed

Commit:
Update Dockerfile

Enable signature checking on pacman

Bobarshad commented on tensorflow/tools/tf_windows_build_dockerfiles/Dockerfile_py3.11:
```diff
@@ -0,0 +1,113 @@
+    Start-Process -FilePath \"C:\Program Files\7-Zip\7z.exe\" -ArgumentList 'x LLVM.exe -oC:\tools\llvm' -Wait; \
+    $env:PATH = [Environment]::GetEnvironmentVariable('PATH', 'Machine') + ';C:\tools\llvm\bin'; \
+    [Environment]::SetEnvironmentVariable('PATH', $env:PATH, 'Machine');
+# Disable signature checking on pacman because we cannot initialize the keyring
```
Enabling signature checking on pacman is done.

Commit:
Update README.md

Commit:
Update Dockerfile

Install msys2 and Visual Studio RUN commands have been moved into a separate script

Commit:
Create msys2.ps1

Commit:
Create visualstudio.bat

Bobarshad committed:
Add files via upload

Moving long RUN commands into the separate scripts

Commit:
Delete visualstudio.bat

angerson commented on tensorflow/tools/tf_windows_build_dockerfiles/bazelisk.ps1:
The line-enders (semicolon, then backtick) shouldn't be necessary in any of these scripts. Please remove them or document why they're there.

Also please add comments on these scripts, at least a header to explain what it's doing.

Bobarshad replied:
When calling the ps1 files in the main Dockerfile, we can use a single RUN command for each script file. By using the backtick character, we can split the command into multiple lines while considering them as a single long command. I will add a header to each sub-script file to explain the purpose of the script.

angerson replied:
I'm pretty sure .ps1 scripts don't need the backticks. Each line gets executed as a separate command.

Commit:
Update 7zip.ps1

Adding header comment

Commit:
Update bazelisk.ps1

Adding header comment

Commit:
Update clang.ps1

Adding header comment.

Commit:
Update jdk.ps1

Adding header comment

Commit:
Update msys2.ps1

Adding header comment

Commit:
Update python.ps1

Adding header comment.

Commit:
Update vs.bat

Adding header comment

Commit:
Update README.md

Adding TF test

Commit:
Update Dockerfile

Enabling long path support

Commit:
Update README.md

adding non-pip test

Commit:
Update README.md

Python config is added for nin-pip test

Commit:
Create docker_win.bazelrc

Commit:
Update README.md

gbaned commented:
Hi @angerson, Can you please review this PR ? Thank you!

gbaned commented:
Hi @angerson, Can you please review this PR ? Thank you!

mihaimaruseac commented:
I'm thinking this can be closed and restarted with the needed changes.

gbaned commented:
Hi @Bobarshad Can you please check @mihaimaruseac 's [comments](https://github.com/tensorflow/tensorflow/pull/60536#issuecomment-1858166038) and keep us posted ? Thank you!

MichaelHudgins commented:
This can be closed for now.  We will be reinvestigating a new container here in the early part of this year.
